access/mattermost/app.go

```diff
@@ -325,57 +412,115 @@ func (a *App) getPostRecipients(ctx context.Context, suggestedReviewers []string
+		return pluginData, true
+	})
+	if err != nil {
 		return trace.Wrap(err)
 	}
-
-	reqData, mmData := pluginData.RequestData, pluginData.MattermostData
-	if len(mmData) == 0 {
-		log.Warn("Failed to update messages. Plugin data is either missing or expired")
+	if !ok {
+		log.Debug("Failed to update posts: plugin data is missing")
 		return nil
 	}
 
-	if err := a.bot.UpdatePosts(ctx, reqID, reqData, mmData, status); err != nil {
+	reqData, mmData := pluginData.RequestData, pluginData.MattermostData
+	if err := a.bot.UpdatePosts(ctx, reqID, reqData, mmData); err != nil {
 		return trace.Wrap(err)
 	}
 
-	log.Infof("Successfully marked request as %s in all messages", status)
+	log.Infof("Successfully marked request as %s in all messages", resolution.Tag)
 
 	return nil
 }
 
-func (a *App) getPluginData(ctx context.Context, reqID string) (PluginData, error) {
-	data, err := a.apiClient.GetPluginData(ctx, types.PluginDataFilter{
+// modifyPluginData performs a compare-and-swap update of access request's plugin data.
+func (a *App) modifyPluginData(ctx context.Context, reqID string, fn func(data *PluginData) (PluginData, bool)) (bool, error) {
+	var lastErr error
+	for i := 0; i < maxModifyPluginDataTries; i++ {
+		oldData, err := a.getPluginData(ctx, reqID)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+		newData, ok := fn(oldData)
+		if !ok {
+			return false, nil
+		}
+		if oldData == nil {
+			err = trace.Wrap(a.setNewPluginData(ctx, reqID, newData))
+		} else {
+			err = trace.Wrap(a.updatePluginData(ctx, reqID, newData, *oldData))
+		}
+		if err == nil {
+			return true, nil
+		}
+		if trace.IsCompareFailed(err) {
+			lastErr = err
+			continue
+		}
+		return false, err
+	}
+	return false, lastErr
+}
+
+// getPluginData loads a plugin data for a given access request. It returns nil if it's not found.
+func (a *App) getPluginData(ctx context.Context, reqID string) (*PluginData, error) {
+	dataMaps, err := a.apiClient.GetPluginData(ctx, types.PluginDataFilter{
 		Kind:     types.KindAccessRequest,
 		Resource: reqID,
 		Plugin:   pluginName,
 	})
 	if err != nil {
-		return PluginData{}, trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
-	if len(data) == 0 {
-		return PluginData{}, nil
+	if len(dataMaps) == 0 {
+		return nil, nil
 	}
-	entry := data[0].Entries()[pluginName]
+	entry := dataMaps[0].Entries()[pluginName]
 	if entry == nil {
-		return PluginData{}, nil
+		return nil, nil
 	}
-	return DecodePluginData(entry.Data), nil
+	data := DecodePluginData(entry.Data)
+	return &data, nil
+}
+
+// setNewPluginData sets plugin data but only if didn't exist before.
+func (a *App) setNewPluginData(ctx context.Context, reqID string, data PluginData) error {
+	expect := make(map[string]string)
+	return a.apiClient.UpdatePluginData(ctx, types.PluginDataUpdateParams{
+		Kind:     types.KindAccessRequest,
+		Resource: reqID,
+		Plugin:   pluginName,
+		Set:      EncodePluginData(data),
+		Expect:   expect,
+	})
 }
```

Comment:
Ah I hadn't update the code yet. After gravitational/teleport#695 I fixed only #220 (pagerduty plugin has the same problem) but forgot to apply the change to slack & mattermost :(

Now I fixed `EncodePluginData` in both slack & mattermost plugins so that if `PluginData` is an empty struct it yields a non-empty map but with all values empty.

For example, before the fix `EncodePluginData` was encoding `0` as a `"0"` now it encodes `0` as `""`